File: C-Programming/Search_Sort/MergeSort.c
#include<stdio.h>
#define SIZE 10

void print(int arr[]);
void mergeSort(int arr[],int base,int end);
void merge(int arr[],int base,int mid, int end);

int main(){
    int arr[SIZE]={5,32,7,2,8,7,9,6,43,65};
    print(arr);
    mergeSort(arr,0,SIZE-1);
    print(arr);

    return 0;
}
void mergeSort(int arr[],int base,int end){
  int mid;
  if (end >base){
    mid = (base+(end-1))/2;
    mergeSort(arr, base, mid);
    mergeSort(arr, mid+1, end);

    merge(arr, base, mid, end);
  }
}
void merge(int arr[], int l,int m, int r){
    int i,j,k,temp;
    i = l;
    j = m+1;

    while(i <= m+2 && j <= r){
        if(arr[i] < arr[j]){
            i++;
        }
        else{
            temp = arr[j];
            for(k=j-1;k>=i;k--)
                arr[k+1] = arr[k];

            arr[i] = temp;
            j++;
        }
    }
}
void print(int arr[]){
    int i;
    printf("Printing Data..: ");

    for(i=0;i<SIZE;i++)
        printf("%d ,",arr[i]);

    printf("\n");
}

File: Python/Algorithm/MergeSort.py

def merge_sort(arr, base, end):
    print(base, end)
    if end > base:
        mid = int((base+(end-1))/2)
        merge_sort(arr, base, mid)
        merge_sort(arr, mid+1, end)
        merge(arr, base, mid, end)


def merge(arr, base, mid, end):
    i = base
    j = mid+1
    while i <= mid and j <= end:
        if arr[i] <= arr[j]:
            i += 1
        else:
            # swap arr[j] with arr[i] and j++
            temp = arr[j]
            for k in range(j,i,-1):
                arr[k] = arr[k-1]
            arr[i] = temp
            j+=1
    # way 2 create a new list add element there in shorted way
    # then overide sorted sublist to our sublist


if __name__ == '__main__':
    data = [34,6,12,446,8,6435,34,4343,768,8,65,446,54,7,68,76]
    merge_sort(data, 0, len(data)-1)
    print(data)
